File: Python3/401 - 600/496 Next Greater Element I.py
# 72ms, 52.18%
class Solution:
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        if not nums1: return []

        res = [-1] * len(nums1)
        num_idx = self.__get_nums_index(nums2)
        rev_max = self.__get_rev_largest_array(nums2, len(nums2))

        for i, n in enumerate(nums1):
            if rev_max[num_idx[n]] > n:
                for j in range(num_idx[n] + 1, len(nums2)):
                    if nums2[j] > n:
                        res[i] = nums2[j]
                        break
        return res

    def __get_nums_index(self, nums):
        return {n: i for i, n in enumerate(nums)}

    def __get_rev_largest_array(self, nums, length):
        ret = [0] * (length - 1) + [nums[-1] - 1]
        for i in range(length - 2, -1, -1):
            ret[i] = max(ret[i + 1], nums[i + 1])
        return ret
    

File: 401 - 600/496 Next Greater Element I.cc
// 4ms, 100.0%
static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
  public:
    vector<int> nextGreaterElement(const vector<int> &findNums, const vector<int> &nums)
    {
        if (findNums.empty()) return vector<int>();
        
        int size1 = findNums.size(), size2 = nums.size();

        unordered_map<int, int> index(size2 * 2);
        for (int i = 0; i < size2; ++i) index[nums[i]] = i;

        vector<int> revMax;
        revMax.resize(size2);
        revMax.back() = nums.back() - 1;
        for (int i = size2 - 2; i >= 0; --i) revMax[i] = max(revMax[i + 1], nums[i + 1]);

        vector<int> res;
        res.resize(size1, -1);
        for (int i = 0; i < size1; ++i) {
            auto idx = index[findNums[i]];
            if (revMax[idx] > findNums[i])
                for (int j = idx + 1; j < size2; ++j)
                    if (nums[j] > findNums[i]) {
                        res[i] = nums[j];
                        break;
                    }
        }
        return res;
    }
};
